javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/InstrumentationModule.java

```diff
@@ -175,6 +179,40 @@ public final AgentBuilder instrument(AgentBuilder parentAgentBuilder) {
     return helperClassNames;
   }
 
+  /**
+   * Returns the matcher produced by {@link #classLoaderMatcher()} with an additional condition
+   * added: no library instrumentation classes defined by this module can be present in the
+   * application classloader. This disables the javaagent instrumentation if the library
+   * instrumentation is already used by the application.
+   */
+  private ElementMatcher.Junction<ClassLoader> createModuleClassLoaderMatcher(
+      List<String> helperClassNames) {
+    // TODO: optimization: refactor matcher caching, only the matcher returned by this method should
+    // cache results, the intermediary ones don't need to cache anything
+    // right now every matcher returned from ClassLoaderMatcher caches separately
+    // this will also make those "Skipping ..." logs appear less
+
+    List<String> libraryHelperClasses =
+        helperClassNames.stream()
+            .filter(this::isLibraryInstrumentationClass)
+            .collect(Collectors.toList());
+
+    if (libraryHelperClasses.isEmpty()) {
+      return classLoaderMatcher();
+    }
+
+    ElementMatcher.Junction<ClassLoader> libraryMatcher =
+        callWhenTrue(
+            hasAnyClassesNamed(libraryHelperClasses), this::logLibraryInstrumentationDetected);
+    return classLoaderMatcher().and(not(libraryMatcher));
+  }
+
+  private void logLibraryInstrumentationDetected() {
```

Comment:
Mmm, if we don't shade library instrumentations, we will detect those classes from within our javaagent jar, no? Or am I asking nonsense?

Reply:
We won't, because they're in the agent classloader.